### PR TITLE
Increment release version just after a release, not just before.

### PR DESCRIPTION
**Pull request by awallace-cray**
This change increments the release version numbers embedded in the
    source, from 1.15.0 to 1.16.0, and adds "pre-release" markings to
    some of those version numbers. Going forward, the "pre-release"
    markings will be maintained on master branch, and removed on each
    new release branch as it is created from master.
    
    The goal is to make intermediate versions of Chapel built from GitHub
    look more like a logical step between releases and not a version of
    the last public release. We also want to clearly mark such versions
    as being "pre-release" versions; i.e., not to be confused with the
    official release to come.

### CODE REVIEW

**Comment by awallace-cray** [comment]
This includes a copy of PR #5962. If this is merged then do not merge that.

**Comment by awallace-cray** [comment]
When I checkout my dev branch from git and run "make", chpl --version gives 
```chpl Version 1.16.0 pre-release (-999)```

When I take a tarball from Jenkins job "realtime-tarball" (built from my dev branch), and run "make" in there, chpl --version gives
```chpl Version 1.16.0 pre-release (4184f5f)```

Is that the expected behavior?

**Comment by awallace-cray** [comment]
If interested in the online docs from this, pls see http://chapel.cray.com/docs/pre-release/index.html

File: CHANGES.md [hunk 1]
```diff
@@ -1,6 +1,12 @@
 Release Changes List
 ====================
 
+version 1.16.0 pre-release
+==========================
+
+*UNDER CONSTRUCTION: Chapel 1.16.0 is scheduled for release in October 2017.*
+
+
```
Comment (bradcray) [inline CHANGES.md]: I wouldn't add this change.  It doesn't really add value, increases the odds that someone will stop reading before getting to the actual valuable content, and is one more thing that will need to be maintained if it's modified now.

File: GOALS.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-Chapel Public Release Goals (version 1.15.0)
+Chapel Public Release Goals (version 1.16.0 pre-release)
```
Comment (bradcray) [inline GOALS.md]: Is it time to just retire the GOALS.md file altogether?  I think of it as more of a "covering our butts" file from the earlier days of the release.  As we become more established and mature, I'd feel less exposed without it.  I/we can do this in a separate PR, though.

**Comment by bradcray** [comment]
I reviewed the files I was most familiar with (primarily those outside of doc/) and other than the minor inline comments, thought this looked good to go.

**Comment by bradcray** [comment]
> When I checkout my dev branch from git and run "make", chpl --version gives
> chpl Version 1.16.0 pre-release (-999)

This occurs when ``$CHPL_HOME`` is unset.  See ``util/devel/updateBuildVersion``

**Comment by bradcray** [comment]
> If interested in the online docs from this, pls see http://chapel.cray.com/docs/pre-release/index.html

I don't understand the relationship between this URL and chapel.cray.com/docs/master (which I also think of as the "pre-release" docs, yet they don't match).

File: doc/rst/conf.py [hunk 3]
```diff
@@ -54,15 +54,15 @@
...
 #
 # The short X.Y version.
-# version = '1.15'
+# version = '1.16'
 
 # We use a custom version variable (shortversion) instead, because setting
 # 'version' adds a redundant version number onto the top of the sidebar
 # automatically (rtd-theme). We also don't use |version| anywhere in rst
-shortversion = 1.15
+shortversion = "1.16 pre-release".replace('-', '&#8209')
```
Comment (ben-albrecht) [inline doc/rst/conf.py]: I would add a comment clarifying the context for this modification:

```python
# Replace `-` with special character `&#8209` to prevent line-breaks in HTML href
```

Comment (ben-albrecht) [inline doc/rst/conf.py]: Also, did you decide against having `conf.py` parse `chpl --version`, or were you planning to do that in a future commit?

**Comment by bradcray** [comment]
> > If interested in the online docs from this, pls see http://chapel.cray.com/docs/pre-release/index.html
> 
> I don't understand the relationship between this URL and chapel.cray.com/docs/master (which I also think of as the "pre-release" docs, yet they don't match).

Thinking about this a few more minutes, I'd really like to see this content be removed from chapel.cray.com ASAP -- it makes me *really* nervous to have new material show up there unanticipated and undiscussed, given that it's the public webspace.  Trials of web-based content should be prototyped and shared within each developer's personal webspace, not pushed to the public website without discussion or review.

**Comment by awallace-cray** [comment]
Okay, there's no more  http://chapel.cray.com/docs/pre-release/index.html

File: doc/rst/conf.py (hunk 3 above)
Comment (awallace-cray) [inline doc/rst/conf.py]: Future. This is just broad strokes.

File: CHANGES.md (hunk 1 above)
Comment (awallace-cray) [inline CHANGES.md]: Okay

File: doc/rst/conf.py (hunk 3 above)
Comment (awallace-cray) [inline doc/rst/conf.py]: I added the comment.

File: doc/rst/conf.py [hunk 4]
```diff
@@ -54,15 +54,17 @@
...
-# version = '1.15'
+# version = '1.16'
 
 # We use a custom version variable (shortversion) instead, because setting
 # 'version' adds a redundant version number onto the top of the sidebar
 # automatically (rtd-theme). We also don't use |version| anywhere in rst
-shortversion = 1.15
+
+shortversion = "1.16 pre-release".replace('-', '&#8209')
+# Replace `-` with special character `&#8209` to prevent line-breaks in HTML href
```
Comment (awallace-cray) [inline doc/rst/conf.py]: I notice one very minor thing about the non-breaking hyphen char. It shows up in the web page "Title".
```<title>Chapel Quickstart Instructions &mdash; Chapel Documentation 1.16 pre&amp;#8209release</title>```
 In Chrome, hover over the Chrome tab w one of these web pages, you can see the  ```&amp;#8209```

This does not seem very important though.

Comment (ben-albrecht) [inline doc/rst/conf.py]: This comes from the line:

```python
html_title = 'Chapel Documentation {0}'.format(shortversion)
```

Perhaps we want something like the following:

```python
chplversion = '1.16 pre-release' # TODO -- parse from `chpl --version`
shortversion = chplversion.replace('-', '&#8209')

...

html_title = 'Chapel Documentation {0}'.format(chplversion)
```

Comment (awallace-cray) [inline doc/rst/conf.py]: Makes sense.

**Comment by awallace-cray** [comment]
Man pages checked. Each man page has ```1.16.0 pre-release``` in the footer. 

I built the tarball with homebrew, and chpl --version gives ```chpl Version 1.16.0 pre-release (5be2e97)```

I intend to make the html_title change that Ben suggested.  Not until we're done reviewing, though. 

Cray module build before merging? (I wasn't planning to, but would do so if asked) 
Anything else?

File: doc/rst/conf.py [hunk 5]
```diff
@@ -54,15 +54,17 @@
...
 
 # The full version, including alpha/beta/rc tags.
-release = '1.15.0'
+release = '1.16.0 pre-release'
```
Comment (ben-albrecht) [inline doc/rst/conf.py]: I think this change is fine, but I'm curious if this shows up anywhere in the generated HTML.. AFAIK, it is not being used currently (I don't see any uses of `|release|` substitution.

File: doc/rst/usingchapel/QUICKSTART.rst [hunk 6]
```diff
@@ -16,7 +20,7 @@ enable more features, such as distributed memory execution.
 0) See :ref:`prereqs.rst <readme-prereqs>` for more information about system
    tools and packages you may need to have installed to build and run Chapel.
 
-1) If you don't already have Chapel 1.15, see
+1) If you don't already have Chapel 1.16, see
```
Comment (ben-albrecht) [inline doc/rst/usingchapel/QUICKSTART.rst]: Note from my earlier [comment](https://github.com/chapel-lang/chapel/issues/5963#issuecomment-292365470) that these should stay pointing towards 1.15 for now, since 1.16 tarball does not exist.

Comment (ben-albrecht) [inline doc/rst/usingchapel/QUICKSTART.rst]: Although, that's just one solution. I'm open to other ideas. My reasoning for the previous approach is that I'm sure users will find their way onto the master-version of the quickstart instructions, and will be thrown off when the 1.16 tarball does not exist...

**Comment by awallace-cray** [comment]
Review comments are adopted. 
https://github.com/chapel-lang/chapel/pull/5992/commits/068233b056c384fe06a757fc486a9f7d909f289c

PR is rebased and re-tested.

**Comment by awallace-cray** [comment]
If you agree, lets merge this. Further improvements can be made later, but this is a good first step

**Comment by ben-albrecht** [comment]
Assuming some sanity checks passed, e.g. building docs, and running any tests that rely on version number, I think it's good to merge.

@awallace-cray -- could you document the known follow-up TODOs in the PR message or #5963 ?

I'd wait for @bradcray's input before merging still.

**Comment by bradcray** [comment]
I stopped tracking this one a few days ago while trying to get release notes done.  If you guys are happy with it and don't have any specific questions you want me to weigh in on, I'd say proceed.

**Comment by awallace-cray** [comment]
@ben-albrecht : TODO list : 
* bradcray - Possibly remove GOALS.md altogether 
* ben-albrecht
  - `doc/rst/conf.py`
    `shortversion = 1.16.0 pre-release` # should be parsed from `chpl --version` if possible
  - `man/conf*.rst`
    It would be nice if the Makefile parsed `chpl --version` and set these automatically
  - `QUICKSTART.rst` & `chplenv.rst`
    To reduce maintenance, we could add a `|lastversion|` sphinx-var defined in `conf.py`, so that references to "release tarball" consistently point to the previous version
